### PR TITLE
Add support for proxy

### LINKED ISSUE #580: Add support to configure an HTTP proxy
Add a HttpHandler to the client options so the user can configure a HTTP proxy

### PR DESCRIPTION
**Pull request by ausfeldt**
## Description

Add support for web proxy and disabling SSL verification in Gateway Mode

```csharp
public IWebProxy CosmosClientOptions.WebProxy;

public CosmosClientBuilder WithConnectionModeGateway(int? maxConnectionLimit = null, IWebProxy webProxy = null);
```

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## Closing issues

closes #580

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs [hunk 1]
```diff
@@ -597,7 +627,28 @@ private void ValidateDirectTCPSettings()
...
+                }
+                else if (this.WebProxy != null)
+                {
+                    settingName = nameof(this.WebProxy);
+                }
+            }
+
+            if (!string.IsNullOrEmpty(settingName))
+            {
+                throw new ArgumentException($"{settingName} requires {nameof(this.ConnectionMode)} to be set to {nameof(ConnectionMode.Gateway)} and {nameof(this.ConnectionProtocol)} to be set to {nameof(Protocol.Https)}");
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs]: ConnectionProtocol is not public contract exclude it from message.

File: Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs [hunk 2]
```diff
@@ -597,7 +627,28 @@ private void ValidateDirectTCPSettings()
             if (!string.IsNullOrEmpty(settingName))
             {
                 throw new ArgumentException($"{settingName} requires {nameof(this.ConnectionMode)} to be set to {nameof(ConnectionMode.Direct)} and {nameof(this.ConnectionProtocol)} to be set to {nameof(Protocol.Tcp)}");
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs]: ConnectionProtocol is not public contract exclude it from message.

**Comment by kirankumarkolli** [comment]
Please add contract changes to the PR description.
Thoughts on a test which can cover a single document CRUD operations through proxy in Gateway mode?

**Comment by kirankumarkolli** [comment]
>public bool CosmosClientOptions.WebProxy;
Need type refresh

**Comment by kirankumarkolli** [comment]
Also can you please add more context how real validation is done?
@Millis mentioned about an issue, lets chat in tomorrow call.

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/CosmosClientOptionsUnitTests.cs [hunk 3]
```diff
@@ -312,13 +320,65 @@ public void VerifyGetConnectionPolicyThrowIfDirectTcpSettingAreUsedInGatewayMode
...
             };
             
             Assert.ThrowsException<ArgumentException>(() => { cosmosClientOptions.IdleTcpConnectionTimeout = idleTcpConnectionTimeout; });
             Assert.ThrowsException<ArgumentException>(() => { cosmosClientOptions.OpenTcpConnectionTimeout = openTcpConnectionTimeout; });
             Assert.ThrowsException<ArgumentException>(() => { cosmosClientOptions.MaxRequestsPerTcpConnection = maxRequestsPerTcpConnection; });
             Assert.ThrowsException<ArgumentException>(() => { cosmosClientOptions.MaxTcpConnectionsPerEndpoint = maxTcpConnectionsPerEndpoint; });
         }
+
+        [TestMethod]
+        public void VerifyHttpClientHandlerSettingsThrowIfNotUsedInGatewayMode()
```
[the remark this answers is not in the document]
Comment (j82w) [reply Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/CosmosClientOptionsUnitTests.cs]: Should we add negative tests to verify the settings changed?

**Comment by kirankumarkolli** [comment]
DisableSSLVerification vs DisableSSLCertificateVerification?

**Comment by ausfeldt** [comment]
> DisableSSLVerification vs DisableSSLCertificateVerification?

DisableSSLVerification is what it is named in the Java SDK. DisableSSLCertificateVerification is better as it is more specific.

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/ClientTests.cs [hunk 4]
```diff
@@ -252,6 +253,23 @@ public void SqlQuerySpecSerializationTest()
...
+        public async Task VerifyNegativeWebProxySettings()
+        {
+            CosmosClient cosmosClient = TestCommon.CreateCosmosClient((cosmosClientBuilder) => {
+                cosmosClientBuilder.WithConnectionModeGateway(webProxy: new TestWebProxy());
+            });
+
+            DatabaseResponse databaseResponse = await cosmosClient.CreateDatabaseAsync(Guid.NewGuid().ToString());
+            Console.WriteLine(databaseResponse.StatusCode);
+
+            /*
```
[the remark this answers is not in the document]
Comment (ausfeldt) [reply Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.EmulatorTests/ClientTests.cs]: it is temporary, trying to figure out why this test passes locally but fails in CI

File: Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs [hunk 5]
```diff
@@ -596,7 +611,24 @@ private void ValidateDirectTCPSettings()
...
+            string settingName = string.Empty;
+            if (this.ConnectionMode != ConnectionMode.Gateway)
+            {
+                if (this.WebProxy != null)
+                {
+                    settingName = nameof(this.WebProxy);
+                }
+            }
+
+            if (!string.IsNullOrEmpty(settingName))
```
[the remark this answers is not in the document]
Comment (ausfeldt) [reply Microsoft.Azure.Cosmos/src/CosmosClientOptions.cs]: there used to be more than one setting. now that it is one I will optimize for one setting.

**Comment by galvesribeiro** [comment]
Hey folks, 

The PR description say:

> Add support for web proxy and disabling SSL verification in Gateway Mode

Where is the support to disable SSL verification? Unfortunately for the lack of cross platform emulator, I have to run the emulator on a VM on OSX, and that makes the SSL to fail. On v2, I was able to add an `HttpClientHandler` to workaround that problem.

Can someone shed a light how this PR would make that workaround to work? I'm quite confuse here...

Thank you!

**Comment by ElasticCoder** [comment]
I agree with @galvesribeiro. I've been working on other things for the last couple of months, but the comment above has led me to look at the SDK again and it appears that we've lost support for disabling SSL verification which we had a couple of months ago?

**Comment by maxdev429** [comment]
@ElasticCoder Is there currently any way to disable SSL verification for CosmosClient in .NET Core?

**Comment by kirankumarkolli** [comment]
@maxdev429 we have a workaround now. Can you please create a new github issue and we can follow-up through it.

/cc: @ealsur (FYR)

**Comment by ealsur** [comment]
@maxdev429 Please see https://docs.microsoft.com/en-us/azure/cosmos-db/local-emulator#running-on-mac-or-linux

You no longer need to disable SSL if you are trying to run the Emulator, exporting the certificate will work on Mac/Linux environment, just make sure you are running the latest emulator.

**Comment by galvesribeiro** [comment]
Yeah, I can confirm that. But be advised that this approach doesn't work you have policies on your machine (i.e. the ones that comes from domain servers) that forbidden untrusted certificates. 

A better way of solving that was if MSFT allowed us to use our own trusted certificate on the emulator.

**Comment by maxdev429** [comment]
@ealsur Specifically I want to automatically run tests using GitLab CI/CD running on Linux and found https://github.com/zeit/cosmosdb-server to emulate a Cosmos database. If I am not mistaken, I would not be able to run the emulate the database using the emulator from the link you posted, as it requires running the emulator from a windows virtual machine, or could I do that using Gitlab CI/CD?

**Comment by ealsur** [comment]
@maxdev429 The Emulator only runs on Windows currently. Please vote for the required feature in the Feedback forums to help prioritize https://feedback.azure.com/forums/263030-azure-cosmos-db/suggestions/18533509-add-documentdb-emulator-support-for-mac-os-x-and

**Pull request merged**